### PR TITLE
Add better handling of JSON-RPC exception

### LINKED ISSUE #36070: Investigate ungraceful failed hot restart request from daemon
Analytics is reporting a high incidence of crashes from the hot restart requests coming from the daemon here: https://github.com/flutter/flutter/blob/v1.7.8%2Bhotfix.3/packages/flutter_tools/lib/src/commands/daemon.dart#L874

As discussed here: https://github.com/flutter/flutter/pull/35846#discussion_r302781867

Please investigate.

So far this is 9.4% of tool crashes in 1.7.8+hotfix.3. This crash was not present in the last stable release.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Fixes https://github.com/flutter/flutter/issues/36070

Catch JSON-RPC exceptions thrown when performing a hot reload or hot restart. Log an analytics event which captures some device specific information. Return a `fatal` result when this happens which informs the tool to exit.

Update TerminalHandler cleanUpAndExit to not perform io.exit.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Good bot!

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -906,8 +909,14 @@ class TerminalHandler {
 
   void registerSignalHandlers() {
     assert(residentRunner.stayResident);
-    io.ProcessSignal.SIGINT.watch().listen(_cleanUpAndExit);
-    io.ProcessSignal.SIGTERM.watch().listen(_cleanUpAndExit);
+    io.ProcessSignal.SIGINT.watch().listen((io.ProcessSignal signal) {
+      _cleanUp(signal);
+      io.exit(0);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Why doesn't throwToolExit work here?

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 2]
```diff
@@ -1051,7 +1066,8 @@ class TerminalHandler {
       await _commonTerminalInputHandler(command);
     } catch (error, st) {
       printError('$error\n$st');
-      await _cleanUpAndExit(null);
+      await _cleanUp(null);
+      rethrow;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: What will catch this? Why not also a throwToolExit?

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 3]
```diff
@@ -946,3 +952,32 @@ class ProjectFileInvalidator {
...
+    } else if (flutterDevices.length > 1) {
+      targetPlatform = 'multiple';
+      deviceSdk = 'multiple';
+      emulator = false;
+    } else {
+      targetPlatform = 'unknown';
+      deviceSdk = 'unknown';
+      emulator = false;
+    }
+    flutterUsage.sendEvent('unhandled_exception', 'hot_mode',
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/run_hot.dart]: 'hot', 'fail' or 'hot', 'exception' would better match existing events, I think.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_runner.dart]: So previously we caught everything and then exited immediately before error reporting. Now we're only catching RpcException and throwing a tool exit in the TerminalHandler.

This rethrow allows us to both rethrow the existing tool exit and handle cleanups so we exit nicely. If there are other exceptions that we're not handling, we should probably let those bubble up so we can fix them.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_runner.dart]: That implies that something went wrong, whereas really this is an exit as normal. I'm also not certain that we have integration tests for these process signals so I'd like to keep it as close to what it was before.

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_hot.dart]: Done

**Pull request merged**